### ISSUE TITLE
Improve collection loading for huge websites

### ISSUE THREAD

**Issue by DavidWells** [MoOx/phenomic#609]
Just tested out a 1000 page site. Generated them with this handy script: https://gist.github.com/DavidWells/a75045c89d6f3d453ef4ab68499550f4

Good news, the build only took like 60 seconds! 

The generated site was 296mb in total though. This was primarily due to every page inlining the entire  `window.__COLLECTION__`  into the html. (https://gist.github.com/DavidWells/675b1b838c44925d6487fff3cc129674)

The `window.__COLLECTION__` makes html files weight 300k vs 8k

Perhaps there is a way to lazy load in the routing data and stick it inside localstorage or a service worker?

Or build the `window.__COLLECTION__` into a single `collection.js` file and include that once in the head tag? Then it could be cached and would make all the html files lighter.

**Comment by thangngoc89** [comment]
Hi, thanks for sugessting this. The collection used to be in a seperated collection.json file and we used to load it async on initial load but it introduces some caveats. 

@MoOx I think dumping it to collection.js and loading it in a blocking way would be the best. (Or non-blocking way if you can fix previous problems)

@DavidWells i tried phenomic on a REAL 10k pages site. The build time is about 8mins and the generated site is 800MB.

**Comment by thangngoc89** [comment]
Well, why did we inject collection to html 

https://github.com/MoOx/phenomic/issues/92

**Comment by MoOx** [comment]
We injected collection for better consitency between client & static build. And for small website, extra http request is not necessary. Obviously for large website it might seems huge, but at the end a single user is supposed to access a single html per session.
We could add a **non lazy** load of the collection via any method, but the collection should be mandatory for the client sync/rendering.

**Comment by DavidWells** [comment]
Maybe there could be an option to inline or include collections via:

```
<script src"/collestions.js?hashXyzydhhs231313123123" async></script>
```

Larger sites could opt for `script` tag and save on overall site size. 

Each inlined collection adds `292kb` to the file size. So an addition 292mb for a site at 1000 pages. 

With the script tag it would be the single 292kb and the deploy/storage would be sooooo much smaller =)

**Comment by MoOx** [comment]
That's definitely something we should do. We could even ask the user during the deployment if we want collection to be outside html (if let say collection > 150kb?) and save the pref in package.json for future build?

**Comment by DavidWells** [comment]
This will make larger sites way more viable and cost friendly

**Comment by DavidWells** [comment]
@MoOx I'd like to take a stab at adding this option.

Would here: https://github.com/MoOx/phenomic/blob/1a11c7f8db4bdc373b72764535f6e5629e042f0a/src/builder/server.js#L137-L141 be the correct place to add the logic to take the collection and add it to script tag instead of inlining?

**Comment by thangngoc89** [comment]
@DavidWells I think it's best to leave it there. It's the dev server. For static build, you should change this line https://github.com/MoOx/phenomic/blob/2b44e19bfa095d2a4bb512c122dbb24380a83ef1/src/static/to-html/url-as-html.js#L108-L111 (extract it to a external js file instead of inline it)

**Comment by MoOx** [comment]
You will also have to emit a collection.json file from Phenomic loader in addition to previous additions you both mentionned.

**Comment by thangngoc89** [comment]
@MoOx it should be collection.js. Let's the browser load that synchronously for us.

**Comment by MoOx** [comment]
Btw, we should replace window.__* by window.__PHENOMIC__.* or something like that... Maybe just window.Phenomic.*

**Comment by thangngoc89** [comment]
I prefer `window.__Phenomic.*` . Who knows what a user can put to the page.

**Comment by MoOx** [comment]
Then, `window.__PHENOMIC_DONT_USE_THIS_OR_YOU_WILL_BE_FIRED__.*`.

More seriously, I will probably use `window.__PHENOMIC_PRIVATE_REGISTRY_YOU_SHOULD_NOT_RELY_ON__*`.

**Comment by thangngoc89** [comment]
@MoOx  

> window.__PHENOMIC_DONT_USE_THIS_OR_YOU_WILL_BE_FIRED__.*.

this is from React code base right ? I like the idea but we referenced this at many places. I don't want to type this everytime

**Comment by MoOx** [comment]
I plan to add public methods like `import { getCollection } from "phenomic"`.

**Comment by thangngoc89** [comment]
Great idea !

**Comment by thangngoc89** [comment]
@DavidWells any update?

**Comment by DavidWells** [comment]
@thangngoc89 how do you think we should implement this?

Are config options available in this context? https://github.com/MoOx/phenomic/blob/2b44e19bfa095d2a4bb512c122dbb24380a83ef1/src/static/to-html/url-as-html.js#L108-L111

I can see a couple ways to turn this on:

- A command line arg `phenomic build --smallHTML(whatever)`
- A config option in package.json
- A webpack config flag?

What do you think?

**Comment by thangngoc89** [comment]
This should be a phenomic's config. FYI all configs in Phenomic can be altered via a cli flag or an entry in package.json/phenomic. 

For adding a new config, see this file https://github.com/MoOx/phenomic/blob/2b44e19bfa095d2a4bb512c122dbb24380a83ef1/src/configurator/definitions.js 

If you need some validation, add one in [validators](https://github.com/MoOx/phenomic/tree/2b44e19bfa095d2a4bb512c122dbb24380a83ef1/src/configurator/validators) folder

**Comment by thangngoc89** [comment]
As for accessing the config, this func https://github.com/MoOx/phenomic/blob/2b44e19bfa095d2a4bb512c122dbb24380a83ef1/src/static/to-html/url-as-html.js#L17 have access to [PhenomicConfig object](https://github.com/MoOx/phenomic/blob/2b44e19bfa095d2a4bb512c122dbb24380a83ef1/interfaces/phenomic.js#L41-L63)

It can be re-writtten like this: `export default (url: String, config: PhenomicConfig, testing?: Boolean)`

**Comment by bloodyowl** [comment]
Going to be fixed with #925
